src/os_pcap_nodelet.cpp

```diff
@@ -0,0 +1,252 @@
+/**
+ * Copyright (c) 2018-2023, Ouster, Inc.
+ * All rights reserved.
+ *
+ * @file os_pcap_nodelet.cpp
+ * @brief This nodelet mainly handles publishing saved metadata
+ *
+ */
+
+// prevent clang-format from altering the location of "ouster_ros/ros.h", the
+// header file needs to be the first include due to PCL_NO_PRECOMPILE flag
+// clang-format off
+#include "ouster_ros/os_ros.h"
+// clang-format on
+
+#include <pluginlib/class_list_macros.h>
+
+#include <string>
+#include <thread>
+
+#include "ouster_ros/os_sensor_nodelet_base.h"
+#include "ouster_ros/PacketMsg.h"
+#include "thread_safe_ring_buffer.h"
+#include <ouster/os_pcap.h>
+
+namespace sensor = ouster::sensor;
+using ouster::sensor_utils::PcapReader;
+
+namespace ouster_ros {
+
+class OusterPcap : public OusterSensorNodeletBase {
+   private:
+    virtual void onInit() override {
+        auto meta_file = get_meta_file();
+        auto pcap_file = get_pcap_file();
+        create_metadata_publisher();
+        load_metadata_from_file(meta_file);
+        allocate_buffers();
+        create_publishers();
+        open_pcap(pcap_file);
+        publish_metadata();
+        create_get_metadata_service();
+        start_packet_processing_threads();
+        start_packet_read_thread();
+        NODELET_INFO("Running in replay mode");
+    }
+
+    std::string get_meta_file() const {
+        auto meta_file =
+            getPrivateNodeHandle().param("metadata", std::string{});
+        if (!is_arg_set(meta_file)) {
+            NODELET_ERROR("Must specify metadata file in replay mode");
+            throw std::runtime_error("metadata no specificed");
+        }
+        return meta_file;
+    }
+
+    std::string get_pcap_file() const {
+        auto pcap_file =
+            getPrivateNodeHandle().param("pcap_file", std::string{});
+        if (!is_arg_set(pcap_file)) {
+            NODELET_ERROR("Must specify pcap file in pcap replay mode");
+            throw std::runtime_error("metadata no specificed");
+        }
+        return pcap_file;
+    }
+
+    void load_metadata_from_file(const std::string& meta_file) {
+        try {
+            cached_metadata = read_text_file(meta_file);
+            info = sensor::parse_metadata(cached_metadata);
+            display_lidar_info(info);
+        } catch (const std::runtime_error& e) {
+            cached_metadata.clear();
+            NODELET_ERROR_STREAM(
+                "Error when running in replay mode: " << e.what());
+        }
+    }
+
+    void allocate_buffers() {
+        auto& pf = sensor::get_format(info);
+
+        lidar_packet.buf.resize(pf.lidar_packet_size);
+        // TODO: gauge necessary queue size for lidar packets
+        lidar_packets =
+            std::make_unique<ThreadSafeRingBuffer>(pf.lidar_packet_size, 1024);
```

Comment:
Is the buffer necessary in this case? Do we really want to start dropping things if ROS cant keep up?

Reply:
I don't see a harm in keeping this queue for now. I will replace with something else in later iterations.

Reply:
Yeah, its just it adds a lot of overhead and complexity.